src/internal/_equals.js

```diff
@@ -1,67 +1,71 @@
+var _arrayFromIterator = require('./_arrayFromIterator');
 var _has = require('./_has');
 var identical = require('../identical');
 var keys = require('../keys');
 var type = require('../type');
 
-// The algorithm used to handle cyclic structures is
-// inspired by underscore's isEqual
-module.exports = function _eqDeep(a, b, stackA, stackB) {
-  var typeA = type(a);
-  if (typeA !== type(b)) {
-    return false;
-  }
-
-  if (typeA === 'Boolean' || typeA === 'Number' || typeA === 'String') {
-    return typeof a === 'object' ?
-      typeof b === 'object' && identical(a.valueOf(), b.valueOf()) :
-      identical(a, b);
-  }
 
+module.exports = function _equals(a, b, stackA, stackB) {
   if (identical(a, b)) {
     return true;
   }
 
-  if (typeA === 'RegExp') {
-    // RegExp equality algorithm: http://stackoverflow.com/a/10776635
-    return (a.source === b.source) &&
-           (a.global === b.global) &&
-           (a.ignoreCase === b.ignoreCase) &&
-           (a.multiline === b.multiline) &&
-           (a.sticky === b.sticky) &&
-           (a.unicode === b.unicode);
+  if (type(a) !== type(b)) {
+    return false;
   }
 
-  if (Object(a) === a) {
-    if (typeA === 'Date' && a.getTime() !== b.getTime()) {
+  switch (type(a)) {
+    case 'Null':
+    case 'Undefined':
       return false;
-    }
-
-    var keysA = keys(a);
-    if (keysA.length !== keys(b).length) {
+    case 'Boolean':
+    case 'Number':
+    case 'String':
+      return typeof a === typeof b && identical(a.valueOf(), b.valueOf());
+    case 'Date':
+      if (!identical(a.valueOf(), b.valueOf())) {
+        return false;
+      }
+      break;
```

Comment:
Why is this branch not simply `return identical(a.valueOf(), b.valueOf());`, you ask? Good question. The test suite currently contains this assertion:

```javascript
var date1 = new Date(0);
var date4 = new Date(0);
date4.v = 'value';
assert.strictEqual(R.equals(date1, date4), false);
```

While this behaviour is reasonable, I can think of no reason it should apply to Date objects but not certain other object types. For example:

```javascript
> var a = /(?:)/
undefined
> var b = /(?:)/
undefined
> b.foo = 'bar'
'bar'
> R.equals(a, b)
true
```

We should either stop checking properties of Date objects or apply these same checks to other object types. This should happen in a separate pull request, but we can discuss the options here.

I lean towards only considering objects equal if they have the same set of properties and the values of each property are equal. It's a good thing, in my view, that `R.equals` does not consider the following values equivalent:

```javascript
> R.match(/foo(bar)?/, 'foobarbaz')
[ 'foobar', 'bar', index: 0, input: 'foobarbaz' ]
> ['foobar', 'bar']
[ 'foobar', 'bar' ]
```